app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/LayoutCalculator.kt

```diff
@@ -31,17 +30,15 @@ data class LayoutCalculator @JvmOverloads constructor(
     }
 
     fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map</* sessionId */ String, LinearLayout.LayoutParams> {
-        val sessions = roomData.sessions
+        val sessions = roomData.sessions.toMutableList()
```

Comment:
Using a mutable list and constantly reading from that list makes it harder than necessary to follow what is going on.

Getting rid of the mutable list could look like this:
```kotlin
    fun calculateLayoutParams(roomData: RoomData, conference: Conference): Map<SessionId, LinearLayout.LayoutParams> {
        val sessions = roomData.sessions
        var previousSessionEndsAt: Int = conference.firstSessionStartsAt.minuteOfDay
        var startTime: Int
        var margin: Int
        var previousSession: Session? = null
        val layoutParamsBySession = mutableMapOf<SessionId, LinearLayout.LayoutParams>()

        for ((index, session) in sessions.withIndex()) {
            startTime = getStartTime(session, previousSessionEndsAt)

            if (startTime > previousSessionEndsAt) {
                // consecutive session
                margin = calculateDisplayDistance(startTime - previousSessionEndsAt)
                if (previousSession != null) {
                    layoutParamsBySession[previousSession.sessionId]!!.bottomMargin = margin
                    margin = 0
                }
            } else {
                // first session
                margin = 0
            }

            val adjustedSession = if (index < sessions.lastIndex) {
                val nextSession = sessions[index + 1]
                fixOverlappingSessions(session, nextSession)
            } else {
                session
            }

            val sessionId = adjustedSession.sessionId
            if (!layoutParamsBySession.containsKey(sessionId)) {
                layoutParamsBySession[sessionId] = createLayoutParams(adjustedSession)
            }

            layoutParamsBySession[sessionId]!!.topMargin = margin
            previousSessionEndsAt = startTime + adjustedSession.duration
            previousSession = adjustedSession
        }

        return layoutParamsBySession
    }

    @VisibleForTesting
    fun fixOverlappingSessions(session: Session, next: Session): Session {
        return if (next.dateUTC > 0 && next.startsAt.isBefore(session.endsAt)) {
            logging.d(LOG_TAG, """Collision: "${session.title}" + "${next.title}"""")
            // cut current at the end, to match next sessions start time
            val newDuration = session.startsAt.minutesUntil(next.startsAt).toInt()
            Session(session).apply { duration = newDuration }
        } else {
            session
        }
    }
```

Reply:
:heavy_check_mark: Thanks a lot for the cleanup. Cute `.withIndex()` - I never used this before.